File: widgetssdk/src/main/java/com/glia/widgets/chat/domain/GliaLoadHistoryUseCase.java
package com.glia.widgets.chat.domain;

import com.glia.androidsdk.chat.ChatMessage;
import com.glia.widgets.chat.data.GliaChatRepository;
import com.glia.widgets.chat.model.history.ChatItem;
import com.glia.widgets.core.engagement.domain.MapOperatorUseCase;
import com.glia.widgets.core.engagement.domain.model.ChatMessageInternal;

import java.util.Comparator;
import java.util.List;

import io.reactivex.Flowable;
import io.reactivex.Single;

public class GliaLoadHistoryUseCase {

    private final GliaChatRepository gliaChatRepository;
    private final MapOperatorUseCase mapOperatorUseCase;

    public GliaLoadHistoryUseCase(GliaChatRepository gliaChatRepository, MapOperatorUseCase mapOperatorUseCase) {
        this.gliaChatRepository = gliaChatRepository;
        this.mapOperatorUseCase = mapOperatorUseCase;
    }

    public Single<List<ChatMessageInternal>> execute() {
        return loadHistory().flatMapPublisher(Flowable::fromArray)
                .concatMapSingle(mapOperatorUseCase::execute)
                .toSortedList(Comparator.comparingLong(o -> o.getChatMessage().getTimestamp()));
    }

    private Single<ChatMessage[]> loadHistory() {
        return Single.create(emitter -> gliaChatRepository.loadHistory((messages, error) -> {
            if (error != null) {
                emitter.onError(error);
            } else {
                emitter.onSuccess(messages);
            }
        }));
    }

}

File: widgetssdk/src/main/java/com/glia/widgets/chat/domain/GliaLoadHistoryUseCase.kt
package com.glia.widgets.chat.domain

import com.glia.widgets.chat.data.GliaChatRepository
import com.glia.widgets.core.engagement.domain.MapOperatorUseCase
import com.glia.widgets.core.engagement.domain.model.ChatHistoryResponse
import com.glia.widgets.core.engagement.domain.model.ChatMessageInternal
import com.glia.widgets.core.secureconversations.SecureConversationsRepository
import com.glia.widgets.core.secureconversations.domain.GetUnreadMessagesCountWithTimeoutUseCase
import com.glia.widgets.core.secureconversations.domain.ManageSecureMessagingStatusUseCase
import io.reactivex.rxjava3.core.Flowable
import io.reactivex.rxjava3.core.Single

internal class GliaLoadHistoryUseCase(
    private val gliaChatRepository: GliaChatRepository,
    private val secureConversationsRepository: SecureConversationsRepository,
    private val shouldUseSecureMessagingApis: ManageSecureMessagingStatusUseCase,
    private val mapOperatorUseCase: MapOperatorUseCase,
    private val getUnreadMessagesCountUseCase: GetUnreadMessagesCountWithTimeoutUseCase
) {

    private val isSecureEngagement get() = shouldUseSecureMessagingApis.shouldUseSecureMessagingEndpoints

    operator fun invoke(): Single<ChatHistoryResponse> = if (isSecureEngagement) {
        loadHistoryWithNewMessagesCount()
    } else {
        loadHistoryAndMapOperator().map { ChatHistoryResponse(it) }
    }

    private fun loadHistoryWithNewMessagesCount() = Single.zip(
        loadHistoryAndMapOperator(),
        getUnreadMessagesCountUseCase()
    ) { messages, count -> ChatHistoryResponse(messages, count) }

    private fun loadHistoryAndMapOperator(): Single<MutableList<ChatMessageInternal>> = loadHistory()
        .flatMapPublisher { Flowable.fromIterable(it) }
        .concatMapSingle { mapOperatorUseCase(chatMessage = it) }
        .toSortedList(Comparator.comparingLong { it.chatMessage.timestamp })

    private fun loadHistory() = Single.create { emitter ->
        loadHistory { messages, error ->
            error?.also { emitter.onError(it) } ?: emitter.onSuccess(messages)
        }
    }

    private fun loadHistory(listener: GliaChatRepository.HistoryLoadedListener) {
        if (isSecureEngagement) {
            secureConversationsRepository.fetchChatTranscript(listener)
        } else {
            gliaChatRepository.loadHistory(listener)
        }
    }
}
